### PR TITLE
Add KeyManager for managing curve key pairs

### PR DESCRIPTION
**Pull request by ankona**

### CODE REVIEW

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py:58]:
```py
private: bytes = dataclasses.field(default=b"", repr=False)
#                                 Good Catch!!  ^^^^^^^^^^
```

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py]: Looks like this is left over. Is it still needed? If not feel free to remove!
```suggestion
        path = self._key_root_dir / self._category
```

File: tests/utils/test_security.py
Comment (MattToast) [inline tests/utils/test_security.py:36]: Pedantic ask for one of these
```py
pytestmark = pytest.mark.group_a
```
or similar so that the CI can pick up these tests.

File: tests/test_config.py
Comment (MattToast) [inline tests/test_config.py]: Style thing if you want to one line this!
```suggestion
    monkeypatch.delenv("SMARTSIM_KEY_PATH", raising=False)
```

File: tests/utils/test_security.py
Comment (MattToast) [reply tests/utils/test_security.py:36]: Also if you could add that whole "BSD 2-Clause" spiel, that would be much appreciated!!

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py]: Just making sure my understanding is correct here: users of this class can set `_separate_keys` by passing in a `seperate_keys={True,False}` parameter in the constructor, but the `_private_subdir` and `_public_subdir` attrs will always be `"priv"` and `"pub"` respectively, right? User of the class cannot (and, in my opinion, at least for now, probably should not be allowed to) alter those subdir names?

If that sounds correct, feel free to resolve! Just want to make sure I'm not wildly off base, lol

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py]: Potentially minor discrepancy here: The description lists `self._category`/`category` as "optional", but it looks like it is a required and non-nullable parameter in the constructor. Can we remove the "(optional)" here?

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py]: From what I can tell, `_client_base` and `_server_base` are only used as constructor args for `_KeyLocator`s for the `_server_locator` and `_client_locator` and attrs respectively, both also instanced in the constructor.

Do these need to be instance attrs? If they do can we mark them `t.Final` just so that they and the `_KeyLocators` don't somehow go "out of sync"?

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py]: Since we have the attrs, can we use
```suggestion
            zmq.auth.create_certificates(locator.private_dir, locator.private.stem)
```
for just a hair more readablility?

File: smartsim/_core/utils/security.py
Comment (MattToast) [inline smartsim/_core/utils/security.py]: If you want, some of these "magic chmod numbers" could be composed from the `stat` constants. For e.g.
```py
stat.S_IRUSR |  stat.S_IWUSR  | stat.S_IXUSR == 0o700 == _KeyPermissions.OWNER_FULL
stat.S_IRUSR |  stat.S_IWUSR                 == 0o600 == _KeyPermissions.OWNER_RW
```
It _might_ be more readable, but honestly that seems like a pretty subjective judgment. I'll leave it to author's discretion!

File: tests/utils/test_security.py
Comment (MattToast) [inline tests/utils/test_security.py]: Hahaha, love the thoroughness!

File: tests/utils/test_security.py
Comment (MattToast) [inline tests/utils/test_security.py]: Pedantic style thing if you want to parameterize these two tests
```py
@pytest.mark.parametrize(
    "as_client, as_server", (
    (     True,      True),
    (     True,     False),
    (    False,      True),
    (    False,     False),
))
def test_key_manager_context(
    test_dir: str, monkeypatch: pytest.MonkeyPatch,
    as_client: bool, as_server: bool,
) -> None:
    """Ensure the key manager does not load private client
    keys when the context is set to server=True"""
    with monkeypatch.context() as ctx:
        ctx.setenv("SMARTSIM_KEY_PATH", test_dir)

        cfg = get_config()
        km = KeyManager(cfg, as_client=as_client, as_server=as_server)

        server_keyset, client_keyset = km.get_keys()

        # as_server=True returns pub/priv server keys...
        assert server_keyset.public
        assert bool(server_keyset.private) == as_server

        # as_server=True returns only public client key
        assert client_keyset.public
        assert bool(client_keyset.private) == as_client
```

File: tests/utils/test_security.py
Comment (MattToast) [inline tests/utils/test_security.py:193]: I'm not sure that I saw a test for the key manager reading keys that already exist. Can we add something like this to the test suite?
```py
def test_key_manager_is_able_to_read_existing_keys_it_created():
    test_dir: str, monkeypatch: pytest.MonkeyPatch
) -> None:
    monkeypatch.setenv("SMARTSIM_KEY_PATH", test_dir)

    cfg = get_config()
    km = KeyManager(cfg, as_client=True, as_server=True)

    # Create some keys
    created_server_keys, created_client_keys = km.get_keys(create=True)
    # ... and then read them
    read_server_keys, read_client_keys = km.get_keys(create=False)
    # ... and check that they are the same
    assert created_server_keys == read_server_keys
    assert created_client_keys == read_client_keys
```

**Review by MattToast — CHANGES_REQUESTED** [review]
A couple of small, mostly overhead-style, tweaks requested, but otherwise LGTM!!

File: smartsim/_core/utils/security.py
Comment (ankona) [reply smartsim/_core/utils/security.py]: Done!

File: tests/utils/test_security.py
Comment (ankona) [reply tests/utils/test_security.py:36]: done. was also missing in `test_dragon_launcher.py` tests so I added it there. Date in `security.py` was updated to 2024

File: tests/test_config.py
Comment (ankona) [reply tests/test_config.py]: nice. Done!

File: smartsim/_core/utils/security.py
Comment (ankona) [reply smartsim/_core/utils/security.py]: Fixed!

File: smartsim/_core/utils/security.py
Comment (ankona) [reply smartsim/_core/utils/security.py]: They got used less & less and finally were totally unnecessary. Great catch. Removed the members and fixed comments

File: smartsim/_core/utils/security.py
Comment (ankona) [reply smartsim/_core/utils/security.py]: Great eye. Swapped to use property instead.

File: smartsim/_core/utils/security.py
Comment (ankona) [reply smartsim/_core/utils/security.py]: Love it. Swapped!

File: tests/utils/test_security.py
Comment (ankona) [reply tests/utils/test_security.py:193]: renamed `test_key_manager_get_existing_keys_only` to  `test_key_manager_get_existing_keys_only_no_keys_found` and added `test_key_manager_get_existing_keys_only_existing` to test this condition. Great suggestion!

File: tests/utils/test_security.py
Comment (ankona) [reply tests/utils/test_security.py]: Made minor modification so I could validate public keys are also loaded, but love it. Done.

File: smartsim/_core/utils/security.py
Comment (ankona) [reply smartsim/_core/utils/security.py]: I definitely didn't want to give customizability to the subdirectories, yet (if ever). I'll argue convention over configurability here is simpler and less hassle for `SmartSim` to fight against user-owned directory permissions.

TBH, I think I don't really need that option to leave keys in the same dir. It actually becomes a headache that way when configuring the zmq `Authenticator` - it wants a directory where public keys exist.

Comment (ankona) [reply smartsim/_core/utils/security.py]: I removed the option. It didn't add anything useful for now.

File: tests/test_dragon_launcher.py
Comment (MattToast) [inline tests/test_dragon_launcher.py]: Looks like a couple of these stub functions are shadowing some previous definitions. While you're here can you resolve them?
- [x] `close`: defined at line 65 and line 71
- [x] `send_json`: defined at line 67 and line 75
- [x] `send_string`: defined at line 69 and line 77

File: tests/test_dragon_launcher.py
Comment (MattToast) [inline tests/test_dragon_launcher.py:38]: Can we delete this block entirely?

**Review by MattToast — APPROVED** [review]
Looks like the only tests that failed were just spontaneous build failures. I found ~two~ one truly pedantic 'clean-up' type thing~s~, but nothing worth holding up approval over!

LGTM!!

File: tests/test_dragon_launcher.py
Comment (MattToast) [reply tests/test_dragon_launcher.py:38]: doh! this was deleted 🤦 ! Ignore me!!

File: tests/test_dragon_launcher.py
Comment (ankona) [reply tests/test_dragon_launcher.py]: derp. i think i put those in 666 and thought i was working in 667. removed

**Review by al-rigazzi — APPROVED** [review]
LGTM! Goodbye eavesdroppers!

**Pull request merged**
